src/sys/mod.rs

```diff
@@ -186,18 +192,46 @@ impl std::fmt::Debug for Poll {
 }
 
 impl Poll {
-    pub(crate) fn new() -> crate::Result<Poll> {
+    pub(crate) fn new(high_precision: bool) -> crate::Result<Poll> {
         Ok(Poll {
-            poller: Poller::new()?,
+            poller: Poller::new(high_precision)?,
             tokens: VecMap::new(),
+            timers: Arc::new(Mutex::new(TimerWheel::new())),
         })
     }
 
     pub(crate) fn poll(
         &mut self,
-        timeout: Option<std::time::Duration>,
+        mut timeout: Option<std::time::Duration>,
     ) -> crate::Result<Vec<PollEvent>> {
-        self.poller.poll(timeout)
+        let now = std::time::Instant::now();
+        // adjust the timeout for the timers
+        if let Some(next_timeout) = self.timers.lock().unwrap().next_deadline() {
+            if next_timeout <= now {
+                timeout = Some(Duration::ZERO);
+            } else if let Some(deadline) = timeout {
+                timeout = Some(std::cmp::min(deadline, next_timeout - now));
+            } else {
+                timeout = Some(next_timeout - now);
+            }
+        };
+
+        let mut events = self.poller.poll(timeout)?;
+
+        let now = std::time::Instant::now();
```

Comment:
There's already a `now` in scope at this point, from L207.

Reply:
Yes, but given we are after `poller.poll()` returns, a lot of time can have elapsed, the point is to update what `now` is. I'll add a comment making this explicit.

Reply:
Oh of course, I missed that.